### PR TITLE
Update prometheus client golang

### PR DESCRIPTION
**Pull request by dashpole**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Update prometheus/client_golang and adapt to behavioral change to the prometheus testutil library.

#### Special notes for your reviewer:

Changes to prometheus tests are is due to [https://github.com/prometheus/client_golang/pull/1424](https://github.com/prometheus/client_golang/pull/1424).  Prior to that change, the metrics were used only as a filter. The test would not fail if a metric in the list was not present. After the change, including a metric that was not present in the exposition would fail the **test.**

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by dashpole** [comment]
https://github.com/prometheus/client_golang/issues/1639 is also related.

cc @bwplotka

**Comment by dims** [comment]
Sigh! on https://github.com/prometheus/client_golang/commit/bccd68204bf41e9cce27e8561b70e0d1cb350bcb

**Comment by kakkoyun** [comment]
> Sigh! on [prometheus/client_golang@bccd682](https://github.com/prometheus/client_golang/commit/bccd68204bf41e9cce27e8561b70e0d1cb350bcb)

@dims, what issues did this commit introduce? It's my mess, and I would like to know more to prevent it from happening again in the future.

**Comment by bwplotka** [comment]
Yea, we can adjust and learn here (e.g. remove a dep)

**Comment by dims** [comment]
> Ok, I see the dependency changes. We will do a better job next time and maybe even remove dependencies.

@kakkoyun thank you! you are right, it's the dependencies. 

When k8s generates all the staging libraries and people use them in the wild, we are passing on these dependencies to them.
- Direct dependencies we pass to those folks will increase their footprint
- Indirect dependencies we pass may sometimes make it harder for them to upgrade/update their existing dependencies.

So we are very very conscious of what we pick up both for ourselves (things that end up in our binaries) but also our libraries that we pass along to our consumers.

**Comment by dims** [comment]
> Yea, we can adjust and learn here (e.g. remove a dep)

@bwplotka thank you! very much appreciate all the work you all do!

File: hack/unwanted-dependencies.json [hunk 1]
```diff
@@ -355,6 +353,7 @@
       "github.com/grpc-ecosystem/go-grpc-prometheus",
       "github.com/grpc-ecosystem/grpc-gateway",
       "github.com/json-iterator/go",
+ 	    "github.com/klauspost/compress",
```
[the remark this answers is not in the document]
Comment (dashpole) [reply hack/unwanted-dependencies.json]: fixed

**Review by logicalhan — COMMENTED** [review]
/lgtm
/approve

**Comment by dashpole** [comment]
/hold

Discussed at sig-instrumentation, and consensus was that we want to explore if a revert of the change in client_golang is possible. @logicalhan @dgrisonnet

**Comment by aramase** [comment]
/lgtm cancel

The changes to the SIG Auth tests (encryption at rest and authenticator) looks unexpected. Please tag @kubernetes/sig-auth-pr-reviews for review after resolving the merge conflicts and https://github.com/kubernetes/kubernetes/pull/127941#issuecomment-2405609184 is resolved.

**Comment by bwplotka** [comment]
Proposed a revert, we will make a decision tomorrow: https://github.com/prometheus/client_golang/pull/1645

**Comment by bwplotka** [comment]
The problematic change was reverted: https://github.com/prometheus/client_golang/releases/tag/v1.20.5

**Comment by dashpole** [comment]
Closing for now.  If needed, i'll open a new PR.

**Pull request closed without merging**